From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Tue, 24 Oct 2017 02:15:13 +0200
Subject: global: make possible to set debug-flags dynamically

Adding {set,get}_debug_flags functions to the shell global object to
make possible to set this easily from looking class, making it easier
for developers and users to debug without having to restart the shell
with environment variables.

Debug flags in main are updated when the "debug-flags" property is
changed. I'm keeping this as a string-list so that it's easier to update.

https://bugzilla.gnome.org/show_bug.cgi?id=789377

Bug-GNOME: https://bugzilla.gnome.org/show_bug.cgi?id=789377
Forwarded: https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/6
---
 src/main.c         | 61 +++++++++++++++++++++++++++++++++++++++++-------------
 src/shell-global.c | 52 ++++++++++++++++++++++++++++++++++++++++++++++
 src/shell-global.h |  4 ++++
 3 files changed, 103 insertions(+), 14 deletions(-)

diff --git a/src/main.c b/src/main.c
index 6895de1..59d78ea 100644
--- a/src/main.c
+++ b/src/main.c
@@ -359,14 +359,14 @@ shell_a11y_init (void)
 }
 
 static void
-shell_init_debug (const char *debug_env)
+shell_update_debug (const char *debug_string)
 {
   static const GDebugKey keys[] = {
     { "backtrace-warnings", SHELL_DEBUG_BACKTRACE_WARNINGS },
     { "backtrace-segfaults", SHELL_DEBUG_BACKTRACE_SEGFAULTS },
   };
 
-  _shell_debug = g_parse_debug_string (debug_env, keys,
+  _shell_debug = g_parse_debug_string (debug_string, keys,
                                        G_N_ELEMENTS (keys));
 }
 
@@ -469,6 +469,42 @@ dump_gjs_stack_on_signal (int signo)
   _tracked_signals[signo] = TRUE;
 }
 
+static void
+reset_signal_handler_to_default (int signo)
+{
+  signal (signo, SIG_DFL);
+  _tracked_signals[signo] = FALSE;
+}
+
+static void
+setup_debug_signal_listners (void)
+{
+  dump_gjs_stack_on_signal (SIGABRT);
+  dump_gjs_stack_on_signal (SIGFPE);
+  dump_gjs_stack_on_signal (SIGIOT);
+  dump_gjs_stack_on_signal (SIGTRAP);
+
+  if ((_shell_debug & SHELL_DEBUG_BACKTRACE_SEGFAULTS))
+    {
+      dump_gjs_stack_on_signal (SIGBUS);
+      dump_gjs_stack_on_signal (SIGSEGV);
+    }
+  else
+    {
+      reset_signal_handler_to_default (SIGBUS);
+      reset_signal_handler_to_default (SIGSEGV);
+    }
+}
+
+static void
+global_notify_debug_flags (GObject    *gobject,
+                           GParamSpec *pspec,
+                           gpointer    data)
+{
+  shell_update_debug (shell_global_get_debug_flags (shell_global_get ()));
+  setup_debug_signal_listners ();
+}
+
 static gboolean
 list_modes (const char  *option_name,
             const char  *value,
@@ -601,6 +637,7 @@ main (int argc, char **argv)
   g_autoptr (GFile) automation_script = NULL;
   g_autofree char *cwd = NULL;
   GError *error = NULL;
+  const char *debug_flags;
   int ecode = EXIT_SUCCESS;
 
   bindtextdomain (GETTEXT_PACKAGE, LOCALEDIR);
@@ -638,17 +675,6 @@ main (int argc, char **argv)
   if (session_mode == NULL)
     session_mode = is_gdm_mode ? (char *)"gdm" : (char *)"user";
 
-  dump_gjs_stack_on_signal (SIGABRT);
-  dump_gjs_stack_on_signal (SIGFPE);
-  dump_gjs_stack_on_signal (SIGIOT);
-  dump_gjs_stack_on_signal (SIGTRAP);
-
-  if ((_shell_debug & SHELL_DEBUG_BACKTRACE_SEGFAULTS))
-    {
-      dump_gjs_stack_on_signal (SIGBUS);
-      dump_gjs_stack_on_signal (SIGSEGV);
-    }
-
   if (script_path)
     automation_script = g_file_new_for_commandline_arg_and_cwd (script_path, cwd);
 
@@ -657,10 +683,16 @@ main (int argc, char **argv)
    * resolve internal modules.
    */
   _shell_global_init ("session-mode", session_mode,
+                      "debug-flags", debug_flags,
                       "force-animations", force_animations,
                       "automation-script", automation_script,
                       NULL);
 
+  g_signal_connect (shell_global_get (), "notify::debug-flags",
+                    G_CALLBACK (global_notify_debug_flags), NULL);
+
+  setup_debug_signal_listners ();
+
   /* Setup Meta _after_ the Shell global to avoid GjsContext
    * iterating on the main loop once Meta starts adding events
    */
@@ -670,7 +702,8 @@ main (int argc, char **argv)
       return EXIT_FAILURE;
     }
 
-  shell_init_debug (g_getenv ("SHELL_DEBUG"));
+  debug_flags = g_getenv ("SHELL_DEBUG");
+  shell_update_debug (debug_flags);
 
   shell_dbus_init ();
   shell_a11y_init ();
diff --git a/src/shell-global.c b/src/shell-global.c
index 4291bb9..69e4838 100644
--- a/src/shell-global.c
+++ b/src/shell-global.c
@@ -61,6 +61,7 @@ struct _ShellGlobal {
   MetaWorkspaceManager *workspace_manager;
 
   char *session_mode;
+  char *debug_flags;
 
   GjsContext *js_context;
   MetaPlugin *plugin;
@@ -119,6 +120,7 @@ enum {
   PROP_SWITCHEROO_CONTROL,
   PROP_FORCE_ANIMATIONS,
   PROP_AUTOMATION_SCRIPT,
+  PROP_DEBUG_FLAGS,
 
   N_PROPS
 };
@@ -235,6 +237,9 @@ shell_global_set_property(GObject         *object,
     case PROP_AUTOMATION_SCRIPT:
       g_set_object (&global->automation_script, g_value_get_object (value));
       break;
+    case PROP_DEBUG_FLAGS:
+      shell_global_set_debug_flags (global, g_value_get_string (value));
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -315,6 +320,9 @@ shell_global_get_property(GObject         *object,
     case PROP_AUTOMATION_SCRIPT:
       g_value_set_object (value, global->automation_script);
       break;
+    case PROP_DEBUG_FLAGS:
+      g_value_set_string (value, global->debug_flags);
+      break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
       break;
@@ -623,6 +631,13 @@ shell_global_class_init (ShellGlobalClass *klass)
                          G_TYPE_FILE,
                          G_PARAM_READWRITE | G_PARAM_CONSTRUCT_ONLY | G_PARAM_STATIC_STRINGS);
 
+  props[PROP_DEBUG_FLAGS] =
+    g_param_spec_string ("debug-flags",
+                         "Debug Flags",
+                         "The debugging flags",
+                         NULL,
+                         G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS);
+
   g_object_class_install_properties (gobject_class, N_PROPS, props);
 }
 
@@ -1767,3 +1782,40 @@ shell_global_set_frame_finish_timestamp (ShellGlobal *global,
       g_object_notify_by_pspec (G_OBJECT (global), props[PROP_FRAME_FINISH_TIMESTAMP]);
     }
 }
+
+/**
+ * shell_global_get_debug_flags:
+ * @global: a #ShellGlobal
+ *
+ * Returns: (transfer none): The current debug flags
+ */
+const gchar *
+shell_global_get_debug_flags (ShellGlobal *global)
+{
+  g_return_val_if_fail (SHELL_IS_GLOBAL (global), NULL);
+
+  return global->debug_flags;
+}
+
+/**
+ * shell_global_set_debug_flags:
+ * @global: a #ShellGlobal
+ * @debug_flags: (nullable): A string for debugging flags
+ *
+ * Updates the debugging flags at runtime as the one set using the SHELL_DEBUG
+ * environment variables. Currently we support 'backtrace-warnings' and
+ * 'backtrace-segfaults' keys.
+ */
+void
+shell_global_set_debug_flags (ShellGlobal  *global,
+                              const char   *debug_flags)
+{
+  g_return_if_fail (SHELL_IS_GLOBAL (global));
+
+  if (g_strcmp0 (global->debug_flags, debug_flags) != 0)
+    {
+      g_free (global->debug_flags);
+      global->debug_flags = g_strdup (debug_flags);
+      g_object_notify (G_OBJECT (global), "debug-flags");
+    }
+}
diff --git a/src/shell-global.h b/src/shell-global.h
index 3bbe3e0..e170ce3 100644
--- a/src/shell-global.h
+++ b/src/shell-global.h
@@ -120,4 +120,8 @@ gboolean shell_global_get_frame_finish_timestamp (ShellGlobal *global);
 void shell_global_set_frame_finish_timestamp (ShellGlobal *global,
                                               gboolean     enable);
 
+const char * shell_global_get_debug_flags       (ShellGlobal  *global);
+void         shell_global_set_debug_flags       (ShellGlobal  *global,
+                                                 const char   *debug_flags);
+
 G_END_DECLS
